woltka/workflow.py

```diff
@@ -0,0 +1,590 @@
+#!/usr/bin/env python3
+
+# ----------------------------------------------------------------------------
+# Copyright (c) 2020--, Qiyun Zhu.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+"""Main classification workflow.
+
+Notes
+-----
+Only in this script can functions directly interface with the user by screen
+output (via `click`) and file input/output, except for raising errors.
+"""
+
+from os import makedirs
+from os.path import join, basename, isfile, isdir
+import click
+
+from .util import (
+    readzip, path2stem, update_dict, allkeys, read_ids, add_dict, intize,
+    delnone, id2file_map, write_map, write_table)
+from .align import Plain, parse_align_file
+from .classify import (
+    assign_none, assign_free, assign_rank, count, strip_index, demultiplex)
+from .tree import (
+    read_names, read_nodes, read_lineage, read_newick, read_ranktb, read_map,
+    fill_root)
+from .ordinal import Ordinal, read_gene_coords, whether_prefix
+
+
+def workflow(input_path, output_path, outmap_dir=None,
+             input_fmt=None, input_ext=None, sample_ids=None, demux=None,
+             ranks=None, above=False, major=None, ambig=True, subok=True,
+             deidx=False, coords_fp=None, overlap=80, names_fp=None,
+             nodes_fp=None, newick_fp=None, lineage_fp=None, ranktb_fp=None,
+             map_fps=[], map_rank=False, lines=1000000):
+    """Main classification workflow which accepts command-line arguments.
+
+    See Also
+    --------
+    .cli.classify
+        Command-line arguments and help information.
+    """
+    # parse input samples
+    samples, files, demux = parse_samples(
+        input_path, input_ext, sample_ids, demux)
+
+    # build classification system
+    tree, rankd, named, root = build_hierarchy(
+        names_fp, nodes_fp, newick_fp, lineage_fp, ranktb_fp, map_fps,
+        map_rank)
+
+    # build processor for alignment
+    proc = build_align_proc(coords_fp, overlap)
+
+    # target classification ranks
+    ranks, rank2dir = prep_ranks(ranks, outmap_dir)
+
+    # classify query sequences
+    data = classify(
+        proc, files, samples, input_fmt, demux, tree, rankd, named, root,
+        ranks, rank2dir, above, major and major / 100, ambig, subok, deidx,
+        lines)
+
+    # write output profiles
+    write_profiles(output_path, data, named, samples)
+    click.echo('Task completed.')
+    return data
+
+
+def classify(proc:    object,
```

Comment:
I think you should instead define something like an abstract class for these "processing" modules. Maybe something like:

```
class AlignmentParser:
    def __init__(*args, **kwargs):
        raise NotImplemented()

    def parse(*args, **kwargs):
        raise NotImplemented()

    def append(*args, **kwargs):
        raise NotImplemented()

    def flush(*args, **kwargs):
        raise NotImplemented()
```
Then this line could be converted to something like:
```suggestion
def classify(proc:    AlignmentParser,
```

Reply:
Also a general comment that I would probably avoid naming things "processor" unless you are dealing with multi-processing as people are bound to confuse "processor" in terms of CPU.

Reply:
I agree with not using "proc", as I also have the same feeling as yours ("proc" means CPU). Meanwhile I cannot use "parser", because this term is already used in `align.py` to indicate functions for parsing individual alignment formats. After bit thinking, I tend to suggest using "**mapper**", because this module basically generates a map between queries and subjects.

I guess the `AlignmentParser` idea may not work. Currently this `proc` (now `mapper`) can be either of two classes: `Plain` (in `align.py`) or `Ordinal` (in `ordinal.py`). They have similar interface hence can be called in the same way, but what they actually do are quite different. So I just leave `object` here.

Also, I tend to downplay Python function hints by not being too specific. More explicit descriptions of parameters are already provided in the docstring, so function hits, which are but hints rather than enforcement, are kind-of redundant. For some big functions I added hints because it looks nice to developers.

Reply:
> Currently this `proc` (now `mapper`) can be either of two classes: `Plain` (in `align.py`) or `Ordinal` (in `ordinal.py`). They have similar interface hence can be called in the same way, but what they actually do are quite different.

You just exactly described why someone might use an abstract class ;)

Reply:
Also "mapper" sounds good!

Reply:
Thanks! @gwarmstrong What is an "abstract class" and is it what you recommend me to use? Sorry I am not familiar with those terms...

Reply:
Here is a python tutorial https://www.geeksforgeeks.org/abstract-classes-in-python/

The tl;dr is that you could name a define a class that does not actually do anything,
but specifies what methods that inherit it should do,

e.g.,

class AlignmentMapper:
    def __init__(*args, **kwargs):
        raise NotImplemented()

    def parse(*args, **kwargs):
        raise NotImplemented()

    def append(*args, **kwargs):
        raise NotImplemented()

    def flush(*args, **kwargs):
        raise NotImplemented()

class OrdinalAlignmentMapper(AlignmentMapper):
    ....

class PlainAlignmentMapper(AlignmentMapper):
    ....
```
That tells you what methods Ordinal and Plain should define, then any time something of AlignmentMapper class is used, you know what behaviors you can expect.

This may not be that important....

What about using `Union` from `typing` in your function definition, so it becomes
```
def classify(proc:    `Union[Plain, Ordinal],
```
So its easy to read but not too much extra work on your end. And `typing` is included in python, no extra dependencies :)

Reply:
Thanks for explaining "abstract class" to me! I read the article you shared, but still am not too clear... but the example you typed looks great! Although I am not sure this exact format, especially `raise NotImplemented()` is what it means to be. I would like to use it if I understand it.

I came across `typing`, thought a bit but didn't use it. That's because if I use it, I will have the tendency to use it in all scenarios, such as `Tuple[str, int]`, and will have to make it very strict. As I explained above, I tend to downplay type hinting since there is already docstring.

Reply:
So `raise NotImplemented()` is an error you can raise to tell whoever called that function that "this has not been implemented yet, please do not use me".

The purpose of `AlignmentMapper` above is to serve as a template to tell people what anything that is an `AlignmentMapper` should have `__init__`, `parse`, `append` and `flush` methods. __But__ you do not actually want to use `AlignmentMapper` since it does not actually do anything, it is just a template. Thus, you want to raise an error so people know they are not supposed to use `AlignmentMapper` directly.

And I agree that type hints are hints, but.... `object` is literally anything in python. Not much of a hint. Even `None` is an `object`! An erratic use of `typing.Union` is preferable to`object`, in my book, as it is much more informative.